File: wpimath/src/main/native/cpp/geometry/proto/QuaternionProto.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <wpi/math/geometry/proto/QuaternionProto.h>

#include "wpimath/protobuf/geometry3d.npb.h"

std::optional<wpi::math::Quaternion>
wpi::Protobuf<wpi::math::Quaternion>::Unpack(InputStream& stream) {
  wpi_proto_ProtobufQuaternion msg;
  if (!stream.Decode(msg)) {
    return {};
  }

  return wpi::math::Quaternion{
      msg.w,
      msg.x,
      msg.y,
      msg.z,
  };
}

bool wpi::Protobuf<wpi::math::Quaternion>::Pack(
    OutputStream& stream, const wpi::math::Quaternion& value) {
  wpi_proto_ProtobufQuaternion msg{
      .w = value.W(),
      .x = value.X(),
      .y = value.Y(),
      .z = value.Z(),
  };
  return stream.Encode(msg);
}

File: wpimath/src/main/java/org/wpilib/math/geometry/proto/QuaternionProto.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.geometry.proto;

import edu.wpi.first.math.geometry.Quaternion;
import edu.wpi.first.math.proto.Geometry3D.ProtobufQuaternion;
import edu.wpi.first.util.protobuf.Protobuf;
import us.hebi.quickbuf.Descriptors.Descriptor;

public class QuaternionProto implements Protobuf<Quaternion, ProtobufQuaternion> {
  @Override
  public Class<Quaternion> getTypeClass() {
    return Quaternion.class;
  }

  @Override
  public Descriptor getDescriptor() {
    return ProtobufQuaternion.getDescriptor();
  }

  @Override
  public ProtobufQuaternion createMessage() {
    return ProtobufQuaternion.newInstance();
  }

  @Override
  public Quaternion unpack(ProtobufQuaternion msg) {
    return new Quaternion(msg.getW(), msg.getX(), msg.getY(), msg.getZ());
  }

  @Override
  public void pack(ProtobufQuaternion msg, Quaternion value) {
    msg.setW(value.getW());
    msg.setX(value.getX());
    msg.setY(value.getY());
    msg.setZ(value.getZ());
  }

  @Override
  public boolean isImmutable() {
    return true;
  }
}
